### PR TITLE
HPCC-22188 Write to FileAccess auditlog using consistent format

### PR DESCRIPTION
**Pull request by wangkx**
The format is: 
```
,FileAccess,<compt-type>,<comp-name>,
  <READ|CREATE|DELETE|COPY|...>,<filepath::filename@nodegroup>,
  <sourcefile::filename@nodegroup>,<user@clientip>,
  "<commadelimitedK=Vstring>"
```

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
@rpastrana please review.

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 1]
```diff
@@ -188,6 +188,14 @@ void CWsDfuEx::init(IPropertyTree *cfg, const char *process, const char *service
     maxFileAccessExpirySeconds = serviceTree->getPropInt("@maxFileAccessExpirySeconds", defaultMaxFileAccessExpirySeconds);
 }
 
+StringBuffer &createAuditUserString(IEspContext &context, const char *userID, StringBuffer &auditUser)
+{
+    if (!isEmptyString(userID))
```
Comment (rpastrana) [inline esp/services/ws_dfu/ws_dfuService.cpp]: for consistency's sake, can we use "ANONYMOUS" or some other value for missing userID

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 2]
```diff
@@ -6571,7 +6581,10 @@ bool CWsDfuEx::onDFUFilePublish(IEspContext &context, IEspDFUFilePublishRequest
 
         newFile->rename(newFileName, userDesc);
 
-        LOG(daliAuditLogCat,",FileAccess,EspProcess,CREATED,%s,%s,%s", groupName, userId.str(), newFileName.str());
+        StringBuffer auditUser;
+        createAuditUserString(context, userId, auditUser);
+        LOG(daliAuditLogCat,",FileAccess,Esp,%s,CREATE,%s@%s,,%s,\"method=WsDfu.DFUFilePublish,LockTimeoutMS=%d\"",
```
Comment (rpastrana) [inline esp/services/ws_dfu/ws_dfuService.cpp]: does the audituser already contain the '@' ip? If not, should it be processed by createAuditUserString()?

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 3]
```diff
@@ -1412,20 +1420,16 @@ bool CWsDfuEx::DFUDeleteFiles(IEspContext &context, IEspDFUArrayActionRequest &r
...
     StringArray superFiles, failedFiles;
 
-    auditStr.append(espProcess.get());
-    auditStr.append(',');
-    if (!isEmptyString(username))
-        auditStr.append(username).append('@');
-    context.getPeer(auditStr);
-
-    doDeleteSuperFiles(req.getLogicalFiles(), userdesc, superFiles, failedFiles, auditStr.str(),
+    VStringBuffer auditStr(",FileAccess,Esp,%s,DELETE", espProcess.get());
```
Comment (rpastrana) [inline esp/services/ws_dfu/ws_dfuService.cpp]: It feels like we should either enumerate the actions (READ|CREATE|DELETE|COPY), or at least #Define them to mitigate possible errors in the future

**Review by rpastrana — CHANGES_REQUESTED** [review]
@wangkx just a few comments

File: esp/services/ws_dfu/ws_dfuService.cpp (hunk 3 above)
Comment (jakesmith) [inline esp/services/ws_dfu/ws_dfuService.cpp]: I think elsewhere audits as "DELETED".. should be consistent.

**Comment by jakesmith** [comment]
@wangkx - I _think_ these changes make the equivalent audit lines in other places (e.g. issues by engines), inconsistent.

For example (from Thor):
```
    LOG(daliAuditLogCat,",FileAccess,EspProcess,READ,%s,%s,%s,jobid=%s,expirySecs=%d", cluster.str(), userID.str(), fileName.str(), requestId, expirySecs);
```

note user is there.

File: esp/services/ws_dfu/ws_dfuService.cpp (hunk 3 above)
Comment (rpastrana) [reply esp/services/ws_dfu/ws_dfuService.cpp]: WE purposely changed the Actions from passed tense to active, but you're right it needs to be changed everywhere.

**Comment by rpastrana** [comment]
@jakesmith this is the structure we designed to be uniform across all components for FileAccess:
,FileAccess,<compt-type>,<comp-name>,
  <READ|CREATE|DELETE|COPY|...>,<filepath::filename@nodegroup>,
  <sourcefile::filename@nodegroup>,<user@clientip>,
  "<commadelimitedK=Vstring>"

File: esp/services/ws_dfu/ws_dfuService.cpp (hunk 3 above)
Comment (jakesmith) [reply esp/services/ws_dfu/ws_dfuService.cpp]: honestly the whole thing should be reworked.. components should not be building these strings up like this manually..
You could do 10% of job now though, I have no objection.

Comment (jakesmith) [reply esp/services/ws_dfu/ws_dfuService.cpp]: Fine, main thing is to be consistent.

File: esp/services/ws_dfu/ws_dfuService.cpp (hunk 2 above)
Comment (jakesmith) [reply esp/services/ws_dfu/ws_dfuService.cpp]: does anywhere contain @ , followed by an IP? The bit after the @ here is a group name..

File: esp/services/ws_dfu/ws_dfuService.cpp (hunk 3 above)
Comment (jakesmith) [reply esp/services/ws_dfu/ws_dfuService.cpp]: Perhaps worth biting the bullet and adding a function and some enums etc. to build up a well formatted/consistent audit log.. 
but that is more work of course and more changes to everywhere else.

Comment (jakesmith) [reply esp/services/ws_dfu/ws_dfuService.cpp]: Perhaps worth biting the bullet and adding a function and some enums etc. to build up a well formatted/consistent audit log.. 
but that is more work of course and more changes to everywhere else.

Comment (rpastrana) [reply esp/services/ws_dfu/ws_dfuService.cpp]: I think I'm 100% in agreement that this should be handled in a centralized function... how are the other components constructing the audit entry?

**Comment by jakesmith** [comment]
right ok, looks fine I think - but need to check everywhere that audits and change accordingly + anywhere that currently reads and processes audit logs.

**Comment by jakesmith** [comment]
><READ|CREATE|DELETE|COPY|...>

Elsewhere (they were elsewhere before added to esp), they were past tense, which is correct, given they are logged after the event.
So should be : READ|CREATED|DELETED|COPIED|..

**Comment by jakesmith** [comment]
Given (to make everything consistent), you will have to change audit logs in many if not all places,
it is worth adding a very simple api/helper (with enumerated types for things like READ/CREATED/DELETED/..) and get everything to use. That way we guarantee everything is consistent now and going forward.

@wangkx back to you.

**Comment by wangkx** [comment]
Ok, I will take a look when I have time.

**Comment by wangkx** [comment]
Close this PR for now. Will reopen it after I finish the fix.

**Comment by rpastrana** [comment]
@wangkx have we had a chance to review this task since the last code review?

**Pull request closed without merging**